File: Python/11.py
# Why didn't I get this problem at first?
    # Brute force is silly
    # Next idea - do left/right maxes like the water fill problem
        # Index only approach not valid since water fills entire interval
    # This algorithm: greedily move smaller boundary
        # Note that there can be scenarios where we need to move the larger
            # boundary to get to an even larger boundary
        # However, this won't improve the filled area until we find a boundary
            # on the other side >= the old boundary we just moved from
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1
        max_area = 0

        while l < r:
            max_area = max(max_area, min(height[l], height[r]) * (r - l))
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        
        return max_area

File: C++/11.cpp
/*
- Notice that greedy approach works better like the 2-sum solution
*/
class Solution {
public:
    int maxArea(vector<int>& height) {
        int i = 0;
        int j = height.size() - 1;
        int cur = 0;
        int out = 0;

        while(i < j){
            cur = (j - i) * min(height[i], height[j]);
            out = max(cur, out);
            if(height[i] <= height[j]){
                i++;
            }
            else{
                j--;
            }
        }
        return out;
    }
};